### PR TITLE
Fix cart pose jacobian calculation and unit test

### PR DESCRIPTION
**Pull request by Levi-Armstrong**

### CODE REVIEW

File: trajopt/src/kinematic_terms.cpp [hunk 1]
```diff
@@ -59,10 +60,7 @@ VectorXd DynamicCartPoseErrCalculator::operator()(const VectorXd& dof_vals) cons
   Eigen::Isometry3d link_tf = world_to_base_ * new_pose * kin_link_->transform * tcp_;
   Eigen::Isometry3d target_tf = world_to_base_ * target_pose * kin_target_->transform * target_tcp_;
 
-  Isometry3d pose_err = target_tf.inverse() * link_tf;
-  Quaterniond q(pose_err.rotation());
-  VectorXd err = concat(Vector3d(q.x(), q.y(), q.z()), pose_err.translation());
-  return err;
+  return calcTransformError(target_tf, link_tf);
```
Comment (mpowelson) [inline trajopt/src/kinematic_terms.cpp]: If you do this, you will need to change the order of the coefficients on this one as well in problem description. You did this for cart pose but I don't think dynamic cart pose was done.

File: trajopt/include/trajopt/utils.hpp [hunk 2]
```diff
@@ -155,4 +156,36 @@ inline std::vector<SafetyMarginData::Ptr> createSafetyMarginDataVector(int num_e
...
+/**
+ * @brief Calculate the rotation error given a rotation error matrix
+ * @param R rotation error matrix
+ * @return rotation error vector
+ */
+inline Eigen::Vector3d calcRotationalError(const Eigen::Ref<const Eigen::Matrix3d>& R)
+{
+  Eigen::AngleAxisd r12(R);
+  double theta = r12.angle();
+  theta = copysign(fmod(fabs(theta),2.0*M_PI), theta);
```
Comment (mpowelson) [inline trajopt/include/trajopt/utils.hpp]: Some comments here would be nice. This gets it into +-2pi and below transforms it into +=pi

File: trajopt/include/trajopt/utils.hpp [hunk 3]
```diff
@@ -155,4 +156,36 @@ inline std::vector<SafetyMarginData::Ptr> createSafetyMarginDataVector(int num_e
   }
   return info;
 }
+
+/**
+ * @brief Calculate the rotation error given a rotation error matrix
```
Comment (mpowelson) [inline trajopt/include/trajopt/utils.hpp]: Add to comment what the error is. ie the axis of the rotation * the magnitude.

**Review by mpowelson — CHANGES_REQUESTED** [review]
The only real change is that I think the coeffs need to be reversed for dynamicCartPose now. We also need to make sure we update the example in Tesseract.

File: trajopt/include/trajopt/utils.hpp (hunk 2 above)
Comment (Levi-Armstrong) [reply trajopt/include/trajopt/utils.hpp]: I read the Eigen source code and they convert the rotation matrix to a quaternion then to angle axis and it states that the angle is between [0, PI] so I will remove this from the code.

**Comment by Levi-Armstrong** [comment]
@mpowelson I have modified the jacobian calculation after reading the following [paper](https://ethz.ch/content/dam/ethz/special-interest/mavt/robotics-n-intelligent-systems/rsl-dam/documents/RobotDynamics2016/RD2016script.pdf). It looks like you can not use the kdl jacobian (Geometric Jacobian) directly. I tried the approach in the paper but need to revisit at another time. I have applied another approach to correctly get the partial derivative of the error function using the Geometric Jacobian. I will work on updating the DynamicCartPose.

**Comment by mpowelson** [comment]
> @mpowelson I have modified the jacobian calculation after reading the following [paper](https://ethz.ch/content/dam/ethz/special-interest/mavt/robotics-n-intelligent-systems/rsl-dam/documents/RobotDynamics2016/RD2016script.pdf). It looks like you can not use the kdl jacobian (Geometric Jacobian) directly. I tried the approach in the paper but need to revisit at another time. I have applied another approach to correctly get the partial derivative of the error function using the Geometric Jacobian. I will work on updating the DynamicCartPose.

For posterity,  Sections 2.8.4 and 2.8.5 are the important ones here.
Analytical: relates differences from joint to configuration space
Geometric: relates the generalized velocity q˙ to the velocity of the end-effector (linear ve and angular ωe):

**Review by mpowelson — APPROVED** [review]

**Pull request merged**